File: leetcode/python/LEET783.py
from typing import *
import sys
import collections
sys.setrecursionlimit(10 ** 9)

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

prev = -sys.maxsize
result = sys.maxsize
def minDiffInBST(root: TreeNode) -> int:
    global result, prev
    if root.left:
        minDiffInBST(root.left)

    result = min(result, root.val-prev)
    prev = root.val

    if root.right:
        minDiffInBST(root.right)
    
    return result

def minDiffInBST2(root: TreeNode) -> int:
    res,  prev = sys.maxsize, -sys.maxsize
    node = root
    stack = []
    while stack or node:
        while node:
            stack.append(node)
            node = node.left
        
        node = stack.pop()
        res = min(res, node.val-prev)
        prev = node.val

        node = node.right
    return res

def solve():
    root = build_tree([27,None,34,None,58,50,None,44])
    res = minDiffInBST(root)
    print(res)

def build_tree(array: List[int]) -> TreeNode:
    if not array:
        return None

    root = TreeNode(array[0])
    q = collections.deque([root])
    index = 1

    while index < len(array):
        node = q.popleft()
        if array[index] is not None:
            left = TreeNode(array[index])
            node.left = left
            q.append(left)
        index += 1
        
        if index >= len(array): break

        if array[index] is not None:
            right = TreeNode(array[index])
            node.right = right
            q.append(right)
        index += 1

    return root

def print_tree(root):
    q = collections.deque([root])

    while q:
        node = q.popleft()
        if node:
            print(node.val, end=" ")
        else:
            print(None, end=" ")
            continue
        
        q.append(node.left)
        q.append(node.right)
    print()
solve()

File: leetcode/java/LEET783.java
import java.util.*;

public class LEET783 {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public static void main(String[] args) {
        TreeNode root = buildTree(new Integer[] {1,0,48,null,null,12,49});
        System.out.println(minDiffInBST(root));
    }
    static int result = 999999999;
    static int prev = -999999999;
    public static int minDiffInBST(TreeNode root) {
        if(root.left != null){
            minDiffInBST(root.left);
        }
        result = Math.min(result, root.val-prev);
        prev = root.val;
        
        if(root.right != null){
            minDiffInBST(root.right);
        }
        return result;
    }
    
    public static int minDiffInBST2(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        int res = 999999999, pre = -99999999;
        TreeNode node = root;

        while(node != null || !stack.isEmpty()){
            while(node != null){
                stack.add(node);
                node = node.left;
            }
            node = stack.pop();
            res = Math.min(res, node.val-pre);
            pre = node.val;

            node = node.right;
        }
        return res;
    }

    public static void printTree(TreeNode root){
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            TreeNode node = q.poll();
            if(node == null){
                System.out.print("NULL ");
                continue;
            }
            System.out.print(node.val +" ");
            
            q.add(node.left);
            q.add(node.right);
        }
        System.out.println();
    }
    public static TreeNode buildTree(Integer[] nums){
        if(nums.length == 0){
            return null;
        }
        Queue<TreeNode> q = new LinkedList<>();
        TreeNode root = new TreeNode(nums[0]);
        int index = 1;
        q.add(root);
        while(index < nums.length){
            TreeNode node = q.poll();
            if(nums[index] != null){
                node.left = new TreeNode(nums[index]);
                q.add(node.left);
            }
            index++;

            if(index >= nums.length)
                break;
            
            if(nums[index] != null){
                node.right = new TreeNode(nums[index]);
                q.add(node.right);
            }
            index++;
        }
        return root;
    }
}
